src/documentation/programming-guide.md

```diff
@@ -1456,4 +1458,202 @@ An example might be if your pipeline reads log records from an input file, and e
+#### Composite Trigger Types
+
+Beam includes the following composite triggers:
+
+*   You can add additional early firings or late firings to `AfterWatermark.pastEndOfWindow` via `.withEarlyFirings` and `.withLateFirings`.
+*   `Repeatedly.forever` specifies a trigger that executes forever. Any time the trigger's conditions are met, it causes a window to emit results and then resets and starts over. It can be useful to combine `Repeatedly.forever` with `.orFinally` to specify a condition that causes the repeating trigger to stop.
+*   `AfterEach.inOrder` combines multiple triggers to fire in a specific sequence. Each time a trigger in the sequence emits a window, the sequence advances to the next trigger.
+*   `AfterFirst` takes multiple triggers and emits the first time *any* of its argument triggers is satisfied. This is equivalent to a logical OR operation for multiple triggers.
+*   `AfterAll` takes multiple triggers and emits when *all* of its argument triggers are satisfied. This is equivalent to a logical AND operation for multiple triggers.
+*   `orFinally` can serve as a final condition to cause any trigger to fire one final time and never fire again.
+
+#### Composition with AfterWatermark.pastEndOfWindow
+
+Some of the most useful composite triggers fire a single time when Beam estimates that all the data has arrived (i.e. when the watermark passes the end of the window) combined with either, or both, of the following:
+
+*   Speculative firings that precede the watermark passing the end of the window to allow faster processing of partial results.
+*   Late firings that happen after the watermark passes the end of the window, to allow for handling late-arriving data
+
+You can express this pattern using `AfterWatermark.pastEndOfWindow`. For example, the following example trigger code fires on the following conditions:
+
+*   On Beam's estimate that all the data has arrived (the watermark passes the end of the window)
+*   Any time late data arrives, after a ten-minute delay
+*   After two days, we assume no more data of interest will arrive, and the trigger stops executing
+
+```java
+  .apply(Window
+      .triggering(AfterWatermark
+           .pastEndOfWindow()
+           .withLateFirings(AfterProcessingTime
+                .pastFirstElementInPane()
+                .plusDelayOf(Duration.standardMinutes(10))))
+      .withAllowedLateness(Duration.standardDays(2)));
+```
+
+#### Other Composite Triggers
+
+You can also build other sorts of composite triggers. The following example code shows a simple composite trigger that fires whenever the pane has at least 100 elements, or after a minute.
+
+```java
+  Repeatedly.forever(AfterFirst.of(
+      AfterPane.elementCountAtLeast(100),
+      AfterProcessingTime.pastFirstElementInPane().plusDelayOf(Duration.standardMinutes(1))))
+```
+
+
+#### Trigger Grammar
+
+The following grammar describes the various ways that you can combine triggers into composite triggers:
+
+```
+TRIGGER ::=
```

Comment:
I'd be tempted to just get rid of the grammar. I don't think it adds much, and each language may realize the grammar differently.